### PR TITLE
Fix DynamicCache reconstruction during ExecuTorch export

### PR DESCRIPTION
**Pull request by eladsegal**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47900)

# What does this PR do?

Fixes incorrect cache-update graphs when exporting models with `DynamicCache` to ExecuTorch.

`torch.export` uses the registered pytree functions to flatten and reconstruct DynamicCache during export. The current pytree registration keeps populated key/value tensors, but loses each layer's type and window size. As a result, updated layers are rebuilt as regular `DynamicLayer`s.

For sliding and chunked attention, `torch.export` can then trace unbounded cache growth instead of the bounded update implemented by `DynamicSlidingWindowLayer`. The immediate logits may still match because the attention mask hides older states, but the cache keeps growing, wasting memory and computation and potentially exceeding ExecuTorch's planned shape or memory limits.

For example, with `sliding_window=4096`, the cache should retain only the latest 4095 past states. Before this change, the exported graph could retain the entire history.

## How it works

The pytree context now records each layer's type, sliding window size, and populated state. Reconstruction recreates the layers in their original positions before restoring their key/value tensors, so `torch.export` traces the correct cache updates.

`DynamicCache` can now be initialized with pre-created layers, allowing reconstruction to establish the fixed layout during initialization. Unsupported specialized cache-layer classes raise a clear error instead of being silently converted.

## Who can review?

@ArthurZucker
@Cyrilvallez

### CODE REVIEW

**Review by ArthurZucker — COMMENTED** [review]
LGTM but @IlyasMoutawwakil for a 2nd look!

**Comment by IlyasMoutawwakil** [comment]
@eladsegal have you tried the exporters api https://huggingface.co/docs/transformers/main/en/exporters ?

**Comment by eladsegal** [comment]
Thanks @IlyasMoutawwakil, I wasn’t familiar with the new exporters API.
I verified that both `DynamoExporter` and `ExecutorchExporter` preserve the layer type, window size, cumulative length and sliding-window cache size limit.
Would you prefer that I close this PR, or update `export_with_dynamic_cache` to delegate to the new exporters API? Or are the helpers in `integrations/executorch.py` intended for deprecation?

**Comment by IlyasMoutawwakil** [comment]
we can merge this one for now, we will deprecate the old ET integration soon once the exporters api is feature complete.

File: src/transformers/cache_utils.py [hunk 1]
```diff
@@ -1776,7 +1779,19 @@ def __init__(
...
     ):
+        if layers is not None:
+            if ddp_cache_data is not None or config is not None:
+                raise ValueError("`layers` cannot be used together with `ddp_cache_data` or `config`.")
+            super().__init__(
+                layers=layers,
+                offloading=offloading,
+                offload_only_non_sliding=offload_only_non_sliding,
+            )
+            return
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py]: Any change here SHOULD NOT be needed. We don't want to update the signature at all. Please use the existing `ddp_cache_data` that will basically do exactly what you want. 
But `cache_utils.py` should NOT be touched here

**Review by Cyrilvallez — COMMENTED** [review]
We cannot accept changes to the signature of the Cache here, especially as what you want to do can be done already through `ddp_cache_data`!

**Comment by Cyrilvallez** [comment]
See [here](https://github.com/huggingface/transformers/blob/main/src/transformers/cache_utils.py#L1789-L1805) it already accepts sliding window

File: src/transformers/integrations/executorch.py [hunk 2]
```diff
@@ -1123,14 +1141,34 @@ def _get_cache_dict(cache: DynamicCache):
...
+    key_states = iter(dictionary.get("key_cache", []))
+    value_states = iter(dictionary.get("value_cache", []))
+    layers = []
+    for sliding_window, is_populated in layout:
+        layer = DynamicLayer() if sliding_window is None else DynamicSlidingWindowLayer(sliding_window)
+        if is_populated:
+            layer.update(next(key_states), next(value_states))
+        layers.append(layer)
+
+    return DynamicCache(layers=layers) if layers else DynamicCache()
```
Comment (IlyasMoutawwakil) [inline src/transformers/integrations/executorch.py]: maybe you can just instantiate the dynamic cache without the layers argument and then manually set the layers attribute to the list built here.

File: src/transformers/cache_utils.py (hunk 1 above)
Comment (eladsegal) [reply src/transformers/cache_utils.py]: Thanks @Cyrilvallez, I removed the `DynamicCache` signature change and now reconstruct the cache through `ddp_cache_data`. I also added a fix to this path to preserve uninitialized layers when both key and value states are `None`.

File: src/transformers/integrations/executorch.py (hunk 2 above)
Comment (eladsegal) [reply src/transformers/integrations/executorch.py]: Thanks @IlyasMoutawwakil, I ended up using the existing `ddp_cache_data` path as Cyril suggested

File: src/transformers/cache_utils.py [hunk 3]
```diff
@@ -1801,8 +1800,12 @@ def __init__(
...
                     else:
                         layers.append(DynamicLayer())
-                # Update the layer with the data
-                _, _ = layers[layer_idx].update(kv_and_optional_sliding[0], kv_and_optional_sliding[1])
+                key_states, value_states = kv_and_optional_sliding[:2]
+                if (key_states is None) != (value_states is None):
+                    raise ValueError("Key and value states must both be tensors or both be `None`.")
+                if key_states is not None and value_states is not None:
+                    # Update the layer with the data
+                    _, _ = layers[layer_idx].update(key_states, value_states)
```
Comment (Cyrilvallez) [inline src/transformers/cache_utils.py:1805]: To remove

**Review by Cyrilvallez — COMMENTED** [review]
Please we want minimal PRs, so revert everything that is not needed and unrelated, especially to general files

File: src/transformers/cache_utils.py (hunk 3 above)
Comment (eladsegal) [reply src/transformers/cache_utils.py:1805]: I added this change to use the `ddp_cache_data` approach you specified.
`export_with_dynamic_cache` uses a cache with layers configured but unpopulated, so unflattening with `ddp_cache_data` has `(None, None, sliding_window)` entries. The existing `DynamicCache.__init__` code calls `layers[layer_idx].update(None, None)` and fails.

This change also aligns the implementation with the existing type annotation:
https://github.com/huggingface/transformers/blob/94f09cfec149050b5355bab7f207ac69e21f1a02/src/transformers/cache_utils.py#L1775

If you prefer to leave `cache_utils.py` untouched, I can instead reconstruct unpopulated layers directly in `integrations/executorch.py` as Ilyas suggested.

Comment (Cyrilvallez) [reply src/transformers/cache_utils.py:1805]: Ha then it's fine, I thought layers would always contain data within executorch, but we can add the small conditional 🤗

**Review by Cyrilvallez — APPROVED** [review]
Alright, pushed a few simplifications, all LGTM now! Thanks a lot! 🤗
